### PR TITLE
Adding Triton implementations awq_dequantize and awq_gemm to ROCm

### PR DESCRIPTION
**Pull request by rasmith**
This PR adds basic support for AWQ by adding Triton implementations of awq_dequantize and awq_gemm to ROCm vLLM.  Performance is not optimized and is quite slow, instead correctness is prioritized.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/envs.py [hunk 1]
```diff
@@ -246,6 +246,10 @@
     # Pad the weight for moe kernel or not
     "VLLM_MOE_PADDING":
     lambda: bool(int(os.getenv("VLLM_MOE_PADDING", "1"))),
+
+    # If set, vllm will print verbose logs during installation
+    "VLLM_USE_TRITON_AWQ":
+    lambda: bool(int(os.getenv("VLLM_USE_TRITON_AWQ", '0'))),
```
Comment (hegemanjw4amd) [inline vllm/envs.py:252]: Can we do this as:

lambda: (os.getenv("VLLM_USE_TRITON_AWQ", "True").lower() in ("true", "1"))

(Make the default "True", to favor Triton.)

File: vllm/model_executor/layers/quantization/awq_triton.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    # NOTE: This doesn't work in TRITON_INTERPRET=1 mode.  Use below instead.
+    # num_pid_n = (N + BLOCK_SIZE_N - 1) // BLOCK_SIZE_N
+    num_pid_n = tl.cdiv(N, BLOCK_SIZE_N)
+
+    pid_m = pid // num_pid_n
+    pid_n = pid % num_pid_n
+
+    accumulator_dtype = c_ptr.type.element_ty
+
+    # NOTE: This doesn't work in TRITON_INTERPRET=1 mode.  Use below instead.
```
Comment (hegemanjw4amd) [inline vllm/model_executor/layers/quantization/awq_triton.py:117]: We'll need to clean this up at some point, but we don't need to do it now.

File: vllm/model_executor/layers/quantization/awq_triton.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+
+@triton.jit
+def awq_dequantize_kernel(
+        qweight_ptr,  # quantized matrix
+        scales_ptr,  # scales, per group
+        zeros_ptr,  # zeros, per group
+        split_k_iters,  # Not used
+        thx,  # Not used
+        thy,  # Not used
```
Comment (hegemanjw4amd) [inline vllm/model_executor/layers/quantization/awq_triton.py:13]: Remove if unused.

File: vllm/model_executor/layers/quantization/awq_triton.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        group_size,  # Should always be 128
```
Comment (hegemanjw4amd) [inline vllm/model_executor/layers/quantization/awq_triton.py:14]: Maybe remove the comment about "always being 128"...  This should work for any reasonable value.

File: vllm/model_executor/layers/quantization/awq_triton.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+# qweights - [K     , M // 8], int32
+# scales   - [K // G, M     ], float16
+# zeros    - [K // G, M // 8], int32
+def awq_dequantize_triton(
+        qweight: torch.Tensor,
+        scales: torch.Tensor,
+        zeros: torch.Tensor,
+        split_k_iters: int,  # Not used
+        thx: int,  # Not used
+        thy: int  # Not used
```
Comment (hegemanjw4amd) [inline vllm/model_executor/layers/quantization/awq_triton.py:230]: Remove.

**Comment by hegemanjw4amd** [comment]
Also, do you want to squash your commits?

File: vllm/envs.py (hunk 1 above)
Comment (rasmith) [reply vllm/envs.py:252]: Oh yes, that makes sense for our repo.

File: vllm/model_executor/layers/quantization/awq_triton.py (hunk 3 above)
Comment (rasmith) [reply vllm/model_executor/layers/quantization/awq_triton.py:13]: Oh yes

File: vllm/model_executor/layers/quantization/awq_triton.py (hunk 4 above)
Comment (rasmith) [reply vllm/model_executor/layers/quantization/awq_triton.py:14]: The group size is set when the model is quantized and it's always 128.  I can update the comment though.

File: vllm/model_executor/layers/quantization/awq_triton.py (hunk 5 above)
Comment (rasmith) [reply vllm/model_executor/layers/quantization/awq_triton.py:230]: Done.

File: vllm/model_executor/layers/quantization/awq_triton.py (hunk 2 above)
Comment (rasmith) [reply vllm/model_executor/layers/quantization/awq_triton.py:114]: Perhaps after optimization is done.  For now, it's very useful.

**Comment by rasmith** [comment]
@hegemanjw4amd If our repo allows, it I'll just squash-merge, which will just result in one single commit for the whole PR.

**Comment by hegemanjw4amd** [comment]
Approving; feel free to fix up minor nits as appropriate.

**Review by hegemanjw4amd — APPROVED** [review]

**Review by shajrawi — APPROVED** [review]
Approving per James' approval

**Pull request merged**